compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -744,15 +744,50 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
                 }
             }
             (Res::Def(DefKind::Struct, def_id), _) if ns == ValueNS => {
-                if let Some((ctor_def, ctor_vis)) = self.r.struct_constructors.get(&def_id).cloned()
+                if let Some((ctor_def, ctor_vis, fields)) =
```

Comment:
You can first obtain the list or private field spans for the `def_id`.
If the list is not empty, then "constructor is not visible here due to private fields" can be reported, otherwise we need to fallback to `bad_struct_syntax_suggestion` below.